vj4/handler/record.py

```diff
@@ -86,3 +91,36 @@ class RecordDetailView(base.Handler):
     rdoc['udoc'], rdoc['pdoc'] = await asyncio.gather(
       user.get_by_uid(rdoc['uid']), problem.get(rdoc['domain_id'], rdoc['pid']))
     self.render('record_detail.html', rdoc=rdoc)
+
+
+@app.route(r'/records/{rid}/pretest_data', 'record_pretest_data')
+class RecordPretestDataView(base.Handler):
+  @base.require_priv(builtin.PRIV_READ_PRETEST_DATA)
+  @base.route_argument
+  @base.sanitize
+  async def get(self, *, rid: objectid.ObjectId):
+    rdoc = await record.get(rid)
+    if not rdoc:
+      raise error.RecordNotFoundError(rid)
+    ddoc = await document.get(rdoc['domain_id'], document.TYPE_PRETEST_DATA, rdoc['data_id'])
+    if not ddoc:
+      raise error.ProblemDataNotFoundError(rdoc['pid'])
+
+    output_buffer = io.BytesIO()
```

Comment:
Currently, the downloader part of judger requires a `content-length` to work, thus zip streaming will not be implemented now. I would like to add a TODO for this.